File: PHP/src/OneLang/Template/Nodes.php
<?php

namespace OneLang\Template\Nodes;

use OneLang\One\Ast\Expressions\Expression;
use OneLang\Utils\TSOverviewGenerator\TSOverviewGenerator;
use OneLang\VM\ExprVM\ExprVM;
use OneLang\VM\ExprVM\VMContext;
use OneLang\VM\Values\ArrayValue;
use OneLang\VM\Values\StringValue;

interface ITemplateNode {
    function format($context);
}

class TemplateBlock implements ITemplateNode {
    public $items;
    
    function __construct($items) {
        $this->items = $items;
    }
    
    function format($context) {
        return implode("", array_map(function ($x) use ($context) { return $x->format($context); }, $this->items));
    }
}

class LiteralNode implements ITemplateNode {
    public $value;
    
    function __construct($value) {
        $this->value = $value;
    }
    
    function format($context) {
        return $this->value;
    }
}

class ExpressionNode implements ITemplateNode {
    public $expr;
    
    function __construct($expr) {
        $this->expr = $expr;
    }
    
    function format($context) {
        $value = (new ExprVM($context))->evaluate($this->expr);
        if ($value instanceof StringValue)
            return $value->value;
        
        if ($context->hooks !== null) {
            $result = $context->hooks->stringifyValue($value);
            if ($result !== null)
                return $result;
        }
        
        throw new \OneLang\Core\Error("ExpressionNode (" . TSOverviewGenerator::$preview->expr($this->expr) . ") return a non-string result!");
    }
}

class ForNode implements ITemplateNode {
    public $variableName;
    public $itemsExpr;
    public $body;
    public $joiner;
    
    function __construct($variableName, $itemsExpr, $body, $joiner) {
        $this->variableName = $variableName;
        $this->itemsExpr = $itemsExpr;
        $this->body = $body;
        $this->joiner = $joiner;
    }
    
    function format($context) {
        $items = (new ExprVM($context))->evaluate($this->itemsExpr);
        if (!($items instanceof ArrayValue))
            throw new \OneLang\Core\Error("ForNode items (" . TSOverviewGenerator::$preview->expr($this->itemsExpr) . ") return a non-array result!");
        
        $result = "";
        foreach (($items)->items as $item) {
            if ($this->joiner !== null && $result !== "")
                $result .= $this->joiner;
            
            $context->model->props[$this->variableName] = $item;
            $result .= $this->body->format($context);
        }
        /* unset @$context->model->props[$this->variableName] ?? null; */
        return $result;
    }
}

File: Python/OneLang/Template/Nodes.py
from onelang_core import *
import OneLang.One.Ast.Expressions as exprs
import OneLang.Utils.TSOverviewGenerator as tSOvervGen
import OneLang.VM.ExprVM as exprVM
import OneLang.VM.Values as vals

class TemplateBlock:
    def __init__(self, items):
        self.items = items
    
    def format(self, context):
        return "".join(list(map(lambda x: x.format(context), self.items)))

class LiteralNode:
    def __init__(self, value):
        self.value = value
    
    def format(self, context):
        return self.value

class ExpressionNode:
    def __init__(self, expr):
        self.expr = expr
    
    def format(self, context):
        value = exprVM.ExprVM(context).evaluate(self.expr)
        if isinstance(value, vals.StringValue):
            return value.value
        
        if context.hooks != None:
            result = context.hooks.stringify_value(value)
            if result != None:
                return result
        
        raise Error(f'''ExpressionNode ({tSOvervGen.TSOverviewGenerator.preview.expr(self.expr)}) return a non-string result!''')

class ForNode:
    def __init__(self, variable_name, items_expr, body, joiner):
        self.variable_name = variable_name
        self.items_expr = items_expr
        self.body = body
        self.joiner = joiner
    
    def format(self, context):
        items = exprVM.ExprVM(context).evaluate(self.items_expr)
        if not (isinstance(items, vals.ArrayValue)):
            raise Error(f'''ForNode items ({tSOvervGen.TSOverviewGenerator.preview.expr(self.items_expr)}) return a non-array result!''')
        
        result = ""
        for item in (items).items:
            if self.joiner != None and result != "":
                result += self.joiner
            
            context.model.props[self.variable_name] = item
            result += self.body.format(context)
        /* unset context.model.props.get(self.variable_name); */
        return result